### PR TITLE
feat(Dropdown): refactor component based on Radix primitives

### PR DESCRIPTION
**Pull request by eszthoff**
BREAKING CHANGE: Dropdown component API and structure changed

Story: [OneUI-496](https://textkernel.atlassian.net/browse/ONEUI-496)

### CODE REVIEW

File: src/components/SelectedItemBadge/SelectedItemBadge.tsx [hunk 1]
```diff
@@ -71,158 +75,123 @@ export interface Props<PriorityItemValue, ChildrenItemValue>
...
-                            {additionalLabel && (
-                                <Text {...elem('optionText')} inline size="small">
-                                    {additionalLabel}
+export const SelectedItemBadge = React.forwardRef<HTMLDivElement, Props<string>>(
+    (
+        {
+            children,
+            label,
+            additionalLabel,
+            // isMultiSelect = false,
```
Comment (AnastasiiaMP) [inline src/components/SelectedItemBadge/SelectedItemBadge.tsx:85]: Could we please update the stories for multiselection of the badge to ensure it works correctly with the new dropdown? Now it says: `Note: group selection and Select all is not correctly implemented in this example. We only showcasing the layout here.`

File: stories/molecules/Dropdown.stories.tsx [hunk 2]
```diff
@@ -2,138 +2,83 @@ import * as React from 'react';
...
 export const _DropdownWithSingleSelectItem: Story = {
     name: 'Dropdown - Single Select Item',
     args: {},
     render: (args) => {
         const [selectedValue, setSelectedValue] = React.useState<string>();
-
-        const onDropdownStateChange = (object) => {
-            console.log(
-                'onDropdownStateChange was called with the following:',
-                JSON.stringify(object)
```
Comment (AnastasiiaMP) [inline stories/molecules/Dropdown.stories.tsx:112]: let’s keep the `console.log` to track state changes, so it stays consistent with all other stories, what do you think?

File: stories/molecules/Dropdown.stories.tsx [hunk 3]
```diff
@@ -211,69 +149,75 @@ export const _DropdownWithMultiSingleSelectItem: Story = {
...
+                                    handleOnChange(value);
+                                }}
+                            >
+                                {value}
+                            </MultiSelectItem>
+                        ))}
+                        <DropdownSub>
+                            <SubContentTrigger>Items →</SubContentTrigger>
+                            <DropdownPortal>
+                                <DropdownSubContent>
```
Comment (AnastasiiaMP) [inline stories/molecules/Dropdown.stories.tsx:203]: The style of the nested field looks different from the design :) 
Screenshot 2024-10-15 at 14 14 08

File: src/components/SelectedItemBadge/SelectedItemBadge.tsx [hunk 4]
```diff
@@ -71,158 +75,123 @@ export interface Props<PriorityItemValue, ChildrenItemValue>
...
+                            <button
+                                aria-label={priority.buttonLabel}
+                                disabled={isDisabled}
+                                type="button"
+                                {...elem('priorityButton')}
+                            >
+                                {renderPriorityIcon(priority.selectedItem?.priority, isDisabled)}
+                            </button>
+                        </DropdownTrigger>
+                        <DropdownContent sideOffset={5}>
```
Comment (AnastasiiaMP) [inline src/components/SelectedItemBadge/SelectedItemBadge.tsx:131]: The `DropdownContent` of the `SelectedItemBadge` must always match the width of the badge and align with it. Previously, we used `refElement={badgeRef}` for this. Currently, it appears misaligned and broken, maybe Radix has something to take care of it 🤔
Screenshot 2024-10-15 at 14 39 57

Screenshot 2024-10-15 at 14 11 32

File: src/components/Dropdown/Items/SingleSelectItem/SingleSelectItem.tsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    id?: string;
+    /** to add a priority badge before the label */
+    hasPriority?: boolean;
+    /** Checkbox status */
+    isSelected?: boolean;
+}
+
+const { block } = bem('DropdownItem', styles);
+
+export const SingleSelectItem = React.forwardRef<HTMLDivElement, Props>(
```
Comment (AnastasiiaMP) [inline src/components/Dropdown/Items/SingleSelectItem/SingleSelectItem.tsx:19]: the cursor of any item (single, multi) looks a little bit confusing:
https://github.com/user-attachments/assets/9919dcf0-1cf5-48bb-96c1-faf16fb9aa83

File: src/components/List/ListItem/index.ts [hunk 6]
```diff
@@ -1,4 +1 @@
 export { ListItem, Props as ListItemProps } from './ListItem';
```
Comment (AnastasiiaMP) [inline src/components/List/ListItem/index.ts:1]: List story is broken now, ones that are not broken looks just weird, but not sure if it is related to the current PR👀
Screenshot 2024-10-15 at 14 50 33

File: stories/organisms/SelectedItemBadge.stories.tsx [hunk 7]
```diff
@@ -132,7 +132,57 @@ export const _SelectedItemBadgeMultiSelect: Story = {
...
+                !selectedSynonyms.includes('All') &&
+                selectedSynonyms.length === synonyms.length + synonyms2.length + 1
+            ) {
+                setSelectedSynonyms([...selectedSynonyms, 'All']);
+            }
+        }, [selectedSynonyms]);
+
+        const handleOnChange = (event: Event, synonym: string) => {
+            // @ts-ignore
+            const checked = !Array.from(event.target.classList).includes(
```
Comment (AnastasiiaMP) [inline stories/organisms/SelectedItemBadge.stories.tsx:153]: do we really need to rely on DOM manipulations? Maybe we could just use selectedSynonyms state to check if it is checked/selected or not? What do you think?

File: stories/organisms/SelectedItemBadge.stories.tsx [hunk 8]
```diff
@@ -161,59 +211,58 @@ export const _SelectedItemBadgeMultiSelect: Story = {
...
+                    onDelete={handleDelete}
+                    additionalLabel={
+                        selectedSynonyms.length > 0 &&
+                        `+${selectedSynonyms.filter((s) => s !== 'All' && s !== 'English').length}`
+                    }
+                >
+                    <MultiSelectItem
+                        variant="select-all"
+                        isSelected={selectedSynonyms.includes('All')}
+                        onSelect={(event) => handleOnChange(event, 'All')}
```
Comment (AnastasiiaMP) [inline stories/organisms/SelectedItemBadge.stories.tsx:233]: `onSelect` closes the dropdown after a single selection, so multi-selection doesn’t work as expected. Also, in `Dropdown` we use `onCheckedChange`, which supports multi-selection, so it’s now unclear which one to use  🤔

File: src/components/Dropdown/Items/MultiSelectItem/MultiSelectItem.tsx [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    DropdownMenuCheckboxItemProps,
+} from '@radix-ui/react-dropdown-menu';
+import { bem } from '../../../../utils';
+import { VisualCheckbox } from '../../../Checkbox';
+import { Text } from '../../../Text';
+import styles from './MultiSelectItem.scss';
+
+export interface Props extends Omit<DropdownMenuCheckboxItemProps, 'onChange'> {
+    /** A function to be called if the item is clicked */
+    onChange?: (e: React.SyntheticEvent<HTMLInputElement | HTMLLIElement>) => void;
```
Comment (AnastasiiaMP) [inline src/components/Dropdown/Items/MultiSelectItem/MultiSelectItem.tsx:13]: we have here `onChange` and `onSelect`, but we use in `MultiSelectItem` stories only `onCheckedChange`, also when we use this component in `Dropdown` story as well, so which one is responsible for what? It's very confusing to me honestly 😅

File: stories/molecules/Dropdown.stories.tsx (hunk 3 above)
Comment (AnastasiiaMP) [reply stories/molecules/Dropdown.stories.tsx:210]: The style still appears incorrect, and it's implemented directly within the stories. Shouldn't we have a dedicated `NestedItem` component as mentioned in the ticket? Or will that be part of a separate ticket?

File: src/components/SelectedItemBadge/SelectedItemBadge.tsx [hunk 10]
```diff
@@ -37,28 +43,24 @@ type PriorityProps<PriorityItemValue> = {
...
+export interface Props<PriorityItemValue>
     extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
     /** Children nodes to be rendered within the Dropdown,
      *  which is triggered by the main button
      * */
     children?:
         | React.ReactElement
-        | React.ReactElement<ListItemProps>
-        | (React.ReactElement<ListItemProps> | EmptyElement)[]
+        | React.ReactElement<SingleSelectItemProps>
```
Comment (AnastasiiaMP) [inline src/components/SelectedItemBadge/SelectedItemBadge.tsx:53]: Maybe we could simplify the children type? before it was so complicated due to the required type in `Dropdown`, now it looks much simpler there. Also, will `SingleSelectItemProps` cover all other options we might pass?

**Review by AnastasiiaMP — CHANGES_REQUESTED** [review]

File: stories/organisms/SelectedItemBadge.stories.tsx (hunk 7 above)
Comment (andrentk) [reply stories/organisms/SelectedItemBadge.stories.tsx:153]: I was not aware of the onCheckedChange

File: stories/organisms/SelectedItemBadge.stories.tsx (hunk 8 above)
Comment (andrentk) [reply stories/organisms/SelectedItemBadge.stories.tsx:233]: onCheckedChange, makes more sense in this case IMO

File: src/components/Dropdown/Items/MultiSelectItem/MultiSelectItem.tsx (hunk 9 above)
Comment (andrentk) [reply src/components/Dropdown/Items/MultiSelectItem/MultiSelectItem.tsx:13]: You're right. I replaced the onChange with the onCheckedChange

File: stories/atoms/MultiSelectItem.stories.tsx [hunk 11]
```diff
@@ -1,9 +1,16 @@
...
+    DropdownRoot,
+    DropdownTrigger,
+    DropdownContent,
+    Button,
+    Text,
+} from '@textkernel/oneui';
 
 export default {
-    title: 'Atoms/List/MultiSelectItem',
+    title: 'Molecules/Dropdown/MultiSelectItem',
```
Comment (eszthoff) [inline stories/atoms/MultiSelectItem.stories.tsx:13]: I think the Dropdown, and all its sub-stories can now move to an Atom, as we have no dependency on other OneUI components, is that correct?

File: stories/atoms/List.stories.tsx [hunk 12]
```diff
@@ -60,8 +59,6 @@ export const _List: Story = {
                 <Checkbox id="item4" />
                 Clickable item
             </ListItem>,
-            <SingleSelectItem>Suggestion Item</SingleSelectItem>,
-            <SingleSelectItem isSelected>Suggestion Item Selected</SingleSelectItem>,
         ],
     },
     render: (args) => <List {...args} />,
```
Comment (eszthoff) [inline stories/atoms/List.stories.tsx:64]: cannot make a comment in the right place, but can you remove https://oneui-git-oneui-496-refactor-68571a-textkernel-oneuis-projects.vercel.app/?path=/story/atoms-list--list-with-separators&vercelToolbarCode=H8DlhrV7xZ6JF0r (separators story)

File: stories/molecules/Dropdown.stories.tsx [hunk 13]
```diff
@@ -2,138 +2,91 @@ import * as React from 'react';
...
-                            <IconTextkernel style={styles.icon} />
-                            <strong>Custom ListItem with value</strong>
-                        </div>
-                    </ListItem>
-                    <div style={styles.customDiv}>Just custom div element</div>
-                </Dropdown>
-            </div>
-        );
-    },
+    render: () => <p>To be implemented</p>,
```
Comment (eszthoff) [inline stories/molecules/Dropdown.stories.tsx:31]: Maybe we should do better then this here. Or if not relevant, let's make a followup ticket for it.

File: stories/molecules/Dropdown.stories.tsx (hunk 2 above)
Comment (eszthoff) [reply stories/molecules/Dropdown.stories.tsx:112]: @AnastasiiaMP what type of `state changes` do you mean? Do we have props for that?

File: stories/molecules/Dropdown.stories.tsx [hunk 14]
```diff
@@ -2,138 +2,91 @@ import * as React from 'react';
...
-                            value={value}
-                            isSelected={selectedValue === value}
+                            onSelect={() => {
+                                handleOnChange('1');
+                            }}
+                            key="1"
+                            isSelected={selectedValue === '1'}
                         >
-                            {value}
+                            Select all
```
Comment (eszthoff) [inline stories/molecules/Dropdown.stories.tsx:62]: I think this value is confusing in the implementation, looks like something you would have in a MultiSelect thing. Also not sure what it supposed to do. Maybe the implementation is not complete...

File: stories/molecules/Dropdown.stories.tsx [hunk 15]
```diff
@@ -211,69 +159,253 @@ export const _DropdownWithMultiSingleSelectItem: Story = {
...
+                        </DropdownSub>
+                    </DropdownContent>
+                </DropdownPortal>
+            </DropdownRoot>
+        );
+    },
+};
+
+export const _DropdownWithSelectedItemBadge: Story = {
+    name: 'Dropdown - SelectedItemBadge',
```
Comment (eszthoff) [inline stories/molecules/Dropdown.stories.tsx:228]: this looks great :-)

File: stories/molecules/Dropdown.stories.tsx (hunk 2 above)
Comment (AnastasiiaMP) [reply stories/molecules/Dropdown.stories.tsx:112]: mostly onChange/onSelect events which changes component’s internal state, such as input values or selected items. Basically as we do in all stories, I am referring to `onDropdownStateChange` here which was removed

**Comment by AnastasiiaMP** [comment]
LGTM 🎉

**Review by AnastasiiaMP — APPROVED** [review]

**Comment by eszthoff** [comment]
LGTM

**Pull request merged**